### PR TITLE
Adding CSRF vulnerability

### PR DESCRIPTION
**Pull request by Zgzorx**
Issue : https://github.com/SasanLabs/VulnerableApp/issues/329

We (@AlexisOsmont  and I) are applying a few changes to the unmerged_csrf branch :
- Using httpOnly and Secure cookies on level 3 and higher
- Using SecureRandom instead of random

Is there anything else needed ?

### CODE REVIEW

File: src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    @VulnerableAppRequestMapping(
+            value = LevelConstants.LEVEL_2,
+            htmlTemplate = "LEVEL_2/CSRFVulnerability",
+            requestMethod = RequestMethod.POST)
+    public ResponseEntity<GenericVulnerabilityResponseBean<String>> getVulnerableCSRFLvl2(
+            @RequestParam(USR_LOGIN) String usrId,
+            @RequestParam(USR_PASS) String usrPass,
+            @RequestParam(USR_NEWPASS) String usrNewPass)
+            throws IOException {
+        return new ResponseEntity<GenericVulnerabilityResponseBean<String>>(
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:122]: please add cookie mechanism here as well.

File: src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            htmlTemplate = "LEVEL_5/CSRFVulnerability",
+            variant = Variant.SECURE)
+    public ResponseEntity<GenericVulnerabilityResponseBean<String>> getVulnerableCSRFLvl5(
+            @RequestParam(USR_LOGIN) String usrId,
+            @RequestParam(USR_PASS) String usrPass,
+            @RequestParam(USR_NEWPASS) String usrNewPass,
+            @RequestParam(TOKEN_PARAM) String usrLiveToken)
+            throws IOException {
+        Map<String, List<String>> headers = new HashMap<>();
+        if (usrLiveToken.equals("secretrequest")) {
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:192]: what is the logic of this if check?

Comment (preetkaran20) [reply src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:192]: I think we can just read the cookie here and see if secret token cookie is there or not and if it there then allow the change of password, else return to login screen.

File: src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            requestMethod = RequestMethod.POST)
+    public ResponseEntity<GenericVulnerabilityResponseBean<String>> getVulnerableCSRFLvl4(
+            @RequestParam(USR_LOGIN) String usrId,
+            @RequestParam(USR_PASS) String usrPass,
+            @RequestParam(USR_NEWPASS) String usrNewPass,
+            @RequestParam(TOKEN_PARAM) String usrLiveToken)
+            throws IOException {
+        Map<String, List<String>> headers = new HashMap<>();
+        headers.put(
+                "Set-Cookie", Arrays.asList("CSRF_Level_4=" + FIXED_TOKEN + "; HttpOnly; Secure"));
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:168]: I think secure flag will not work as we are not hosting application on https.

File: src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+            description = "CSRF_VULNERABILITY_USER_INPUT1",
+            payload = "CSRF_PAYLOAD_LEVEL_3")
+    @VulnerableAppRequestMapping(
+            value = LevelConstants.LEVEL_3,
+            htmlTemplate = "LEVEL_3/CSRFVulnerability")
+    public ResponseEntity<GenericVulnerabilityResponseBean<String>> getVulnerableCSRFLvl3(
+            @RequestParam(USR_LOGIN) String usrId,
+            @RequestParam(USR_PASS) String usrPass,
+            @RequestParam(USR_NEWPASS) String usrNewPass,
+            @RequestParam(TOKEN_PARAM) String usrLiveToken)
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:139]: this param is not required.

File: src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    @AttackVector(
+            vulnerabilityExposed = VulnerabilityType.CSRF,
+            description = "CSRF_VULNERABILITY_USER_INPUT1",
+            payload = "CSRF_PAYLOAD_LEVEL_1")
+    @VulnerableAppRequestMapping(
+            value = LevelConstants.LEVEL_1,
+            htmlTemplate = "LEVEL_1/CSRFVulnerability")
+    public ResponseEntity<GenericVulnerabilityResponseBean<String>> getVulnerableCSRFLvl1(
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:98]: I would suggest that create another Api for just login and let this one for only change password.

So one method per vulnerability will be for login and that will be first called by the user. once login is completed, then show change password UI screen and let user change the password.

File: src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    @AttackVector(
+            vulnerabilityExposed = VulnerabilityType.CSRF,
+            description = "CSRF_VULNERABILITY_USER_INPUT2",
+            payload = "CSRF_PAYLOAD_LEVEL_5")
+    @VulnerableAppRequestMapping(
+            value = LevelConstants.LEVEL_6,
+            htmlTemplate = "LEVEL_6/CSRFVulnerability",
+            variant = Variant.SECURE,
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:220]: this is not a secure level, i would suggest going through https://www.youtube.com/watch?v=hW2ONyxAySY&ab_channel=TroyHunt video for information about CSRF.

File: src/main/java/org/sasanlabs/service/vulnerability/sampleVulnerability/SampleVulnerability.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+         * as the description of the Vulnerability. It helps students to learn about the
+         * vulnerability and can also include some of the useful references etc.
+         */
+        descriptionLabel = "SAMPLE_VULNERABILITY",
+        /**
+         * "value" parameter of annotation is used to create the request mapping. e.g. for the below
+         * parameter value, /VulnerableApp/SampleVulnerability will be created as URI Path.
+         */
+        value = "SampleVulnerability")
+public class SampleVulnerability {
```
Comment (preetkaran20) [inline src/main/java/org/sasanlabs/service/vulnerability/sampleVulnerability/SampleVulnerability.java:35]: please remove this file. this file should not be there.

File: src/main/resources/static/templates/SampleVulnerability/LEVEL_1/SampleVulnerability.css [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#SampleVulnerability {
```
Comment (preetkaran20) [inline src/main/resources/static/templates/SampleVulnerability/LEVEL_1/SampleVulnerability.css:1]: please remove this file

File: src/main/resources/static/templates/SampleVulnerability/LEVEL_1/SampleVulnerability.html [hunk 9]
```diff
@@ -0,0 +1,9 @@
+<div id="SampleVulnerability">
+	<div>
+		<div id="level_info">
```
Comment (preetkaran20) [inline src/main/resources/static/templates/SampleVulnerability/LEVEL_1/SampleVulnerability.html:3]: this one as well.

File: src/main/resources/static/templates/SampleVulnerability/LEVEL_1/SampleVulnerability.js [hunk 10]
```diff
@@ -0,0 +1,23 @@
+function addingEventListenerToFetchData() {
+  document
```
Comment (preetkaran20) [inline src/main/resources/static/templates/SampleVulnerability/LEVEL_1/SampleVulnerability.js:2]: this one too.

File: src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java (hunk 6 above)
Comment (preetkaran20) [reply src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:220]: and https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html for mitigating this risk in applications.

File: src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java (hunk 2 above)
Comment (Zgzorx) [reply src/main/java/org/sasanlabs/service/vulnerability/csrf/CSRFVulnerability.java:194]: I am not sure about this, it was done by the original author of the file. I'll dig into it

**Pull request closed without merging**